src/condor_utils/file_transfer.cpp

```diff
@@ -6761,6 +6852,56 @@ FileTransfer::InvokeMultipleFileTransferPlugin( CondorError &e, int &exit_status
 		}
 	}
 
+
+	//
+	// Create a PluginInvocation record each time we invoke a plug-in.  We
+	// fill it out as we go along, and the convert it into a ClassAd so it
+	// can be sent to the shadow with the individual URL result ads.
+	//
+	PluginInvocation pi;
+
+	pi.transfer_class = TransferClass::input;
+	if( do_upload ) {
+		pi.transfer_class = TransferClass::output;
+		if( uploadCheckpointFiles ) {
+			pi.transfer_class = TransferClass::checkpoint;
+		}
+	}
+
+	// Arguably, this should just be `plugin.name`.
+	pi.plugin_basename = condor_basename(plugin.path.c_str());
+
+
+	// The parameter transfer_files_string is a series of two-element
+	// ClassAds in new ClassAd form without any separators.  The two
+	// attributes are `LocalFileName`, which we don't care about, and `URL`,
+	// which we'd like to parse for its schema.
+	//
+	// Unfortunately, StringTokenIterator doesn't accept multicharacter
+	// delimiters, so we have to do this the hard way.
+	size_t equals = 0;
+
+	while( true ) {
+		equals = transfer_files_string.find( "Url = \"", equals );
+		if( equals != std::string::npos ) {
+			size_t start_of_schema = equals + 7;
+			size_t end_of_schema = transfer_files_string.find( "://", start_of_schema );
+			if( end_of_schema != std::string::npos ) {
+				std::string schema = transfer_files_string.substr( start_of_schema, end_of_schema - start_of_schema );
+				pi.schemes.push_back( schema );
+
+				equals = end_of_schema;
+			} else {
+				break;
+			}
+		} else {
+			break;
+		}
+	}
```

Comment:
We should explore how this code interacts with token/service specification on the schemas? If the FTO is passing the full `service+schema://` to the plugin then this list is not what we want.

Reply:
The FTO calls getURLType(, TRUE) to extract the "method" ("schema") from a URL; this function should do the same.